### PR TITLE
Add datatype(s): RDS object and SingleCellExperiment class

### PR DESCRIPTION
**Pull request by mtekman**
## General

This PR includes a new datatype RDS that is used to serialize/de-serialize R objects.

The RDS datatype can also be used to sniff for specific class instances, in this case for the SingleCellExperiment class.

## Motivation

The idea behind implementing these types is to allow some flexibility in generating single cell RNA (scRNA)  downstream analysis workflows. 

The development of scRNA tools resides primarily in R; an unfortunate choice of environment, but one that is unlikely to change.

Most R tools are adopting the SingleCellExperiment object to perform most of their analyses which consist of 4 key stages:

  1. Filtering
  2. Normalisation
  3. Confounder removal
  4. Clustering

As I see it, there are two possible methods of catering for this problem:
 
 1. Keep all downstream pipelines contained as standalone tools, removing the need to deal with filetypes. This would mean that the four standard analysis processes described above would all be performed within the same (extensively convoluted) tool. Pipelines that differ only in small specific areas would still have to be reproduced entirely.

 2. Separate the four stages into their own tools, and join them together like any other workflow. This would allow more choice for the user to decide *which* filtering tool they wish to or *which* normalisation tool they wish to use, etc.  Tools can flow into each other like any other workflow, and users would be free to construct their own pipelines.

### CODE REVIEW

File: config/datatypes_conf.xml.sample [hunk 1]
```diff
@@ -524,6 +524,7 @@
     <datatype extension="xmfa" type="galaxy.datatypes.msa:MauveXmfa" display_in_upload="true" />
     <datatype extension="cel" type="galaxy.datatypes.binary:Cel" display_in_upload="true" />
     <datatype extension="rdata" type="galaxy.datatypes.binary:RData" display_in_upload="true" description="Stored data from an R session"/>
+    <datatype extension="sce" type="galaxy.datatypes.binary:RData" description="Stored RDS from a SingleCellObject" subclass="true" />
```
Comment (bgruening) [inline config/datatypes_conf.xml.sample]: should we name it `rdata.sce`?

Comment (mtekman) [reply config/datatypes_conf.xml.sample]: sounds good!

**Comment by jmchilton** [comment]
Pushed a merge commit into your branch to resolve the conflict that had been introduced, hope that is okay. This looks good to me - thanks for the contribution @mtekman!

**Comment by mblue9** [comment]
Hello, I'm doing some work on the seurat tool and wondering if the datatype added here rdata.sce could be made more generic to accept other rds types e.g the seurat object? 
Also wondering if the datatype needs to be specific to single cell or could it simply be rds to allow any R rds object?

**Comment by mtekman** [comment]
I saw, nice stuff! 

I'm not sure how generic it can be. I figure as long as all the slots are the same, then there should be no problem. Does Seurat use the SCE base class?

**Comment by mtekman** [comment]
It looks like there is [some compatibility](https://satijalab.org/seurat/conversion_vignette.html), but the base class is their own.

I think it would be better to extend the rdata class to `rdata.seurat`. If I recall, I made the RDS class generic enough to test for the type of RDS by `class(obj)`

**Comment by mblue9** [comment]
Thanks for the quick reply @mtekman ! And thanks for the info! What would I need to do to extend the rdata class to `rdata.seurat` (and where) ?

**Comment by mtekman** [comment]
Ah, I just realized I had to undo a lot of previous edits, and that the class was never implemented

https://github.com/galaxyproject/galaxy/pull/5716/commits/e89290c33031d89365aa3ef3b6e4720d68390119#diff-409951594551183dbf886e24de6cb129

Perhaps it is now the time to re-implement this?  @bgruening

**Comment by mblue9** [comment]
I was wondering why those commits were removed :) Ok, let me know if I need to do something here.

**Comment by bgruening** [comment]
Sure, feel free :)

**Comment by mtekman** [comment]
If I recall correctly, the reasoning was that if these data types were
limited to a single tool (at the time just Scater) then it would create
needless overhead during file type sniffing, since only a single tool would
have that file format.

So we abandoned the new filetype class, and instead just extended the
existing rdata data type.

This worked well enough, and I think could also possibly still work here
where you could add a single line in the datatypes file with your own
Seurat type which also extends rdata (as shown in the very final commit) -
but to me it would make more sense if they inherited from the RDS class I
had previously tried to implement, because RData and RDS are two very
different things in the R ecosystem.

Essentially, I dont know - what do you think is best? :-)

If you want a fast solution, then just replicate the single line in my last
commit

**Comment by mblue9** [comment]
Thanks @mtekman! For the moment I've gone with the fast solution here https://github.com/galaxyproject/galaxy/pull/6648 but it could make sense to add an RDS class at some point.

**Comment by mblue9** [comment]
Out of interest are these file type classes only used when auto-detecting file types during upload?  As if so, then just to say, personally I never use auto-detect anymore, I always specify the file type. So my preference would be just to have a single RData and RDS file type format (as in R) and to drop the specific formats rdata.sce and rdata.seurat, as am wondering how many more rdata.x's (or rds.x's) this current method could end up resulting in.

**Pull request merged**